### ISSUE TITLE
Question: how to find all GEP reads and writes to a specific variable?

### ISSUE THREAD

**Issue by enzo-pellegrini** [SVF-tools/SVF#1243]
Hi, I am trying to find all reads and writes to fields of a variable (struct or array). This is because I want to instrument the corresponding LLVM IR to behave differently (the allocations are changed so that some of the variables in a program are moved to a webassembly memory, and so the pointers have to actually be computed differently to be 8 bytes wide, just to give a bit of context).

So just a simple example.

```c
struct s {
  int *a;
};

int main() {
  struct s* p = malloc(sizeof(struct s)); // this gets found through points-to analysis and changed
  int *n = malloc(sizeof(int)); // this gets found through points-to analysis and changed.
  p->a = n; // I want to get a pointer to the gep that is part of the translation of this line and also to the reads and writes done on the resulting pointer 
  interesting_f(p);
}
```

So I'm already able to find the allocation of n (because I change that too) using getAllFieldsObjVar, but what api should I use to find the GEP instruction that computes the pointer to field a of p (and then all writes and reads to it I guess should be available from pointed to)? Should I use the Value Flow Graph (can I get from that a pointer to the LLVM instruction?) or can I use the constraint graph?

So tldr:
- Is the constraint graph supposed to be something that I can use for value flow analysis (in this case, finding this GEP instruction that gets a pointer to a field of structs and arrays I'm interested in) or is something to be used only internally or to build other points-to analysis?
- Can I go from a Value Flow Graph node to the underlying SVF instruction (in this case the GEP)? Edit ok I see that there's a getValue method on VFGNode and then I can maybe check this and the node type by getNodeKind() to check if it's a GEP.

Anyway sorry for the long question, I was a bit confused from the docs but I am very grateful that a library like this exists and is available to everyone. Hope you can give some pointers on which api or graph I'm supposed to explore.

**Comment by yuleisui** [comment]
From my understanding, a simple solution is to run Andersen’s or flow-sensitive analysis. For each objects you collected, get their reverse points-to (getRevPts), then you will get all pointers that access this object. You may only be interested in pointers are operants at stores, loads and getelementptrs.

**Comment by yuleisui** [comment]
You could also query getPts for any pointer at a load/store of your interest. If the points to set contains the object you collected. The pointer at that load/store would be something you wanted

**Comment by enzo-pellegrini** [comment]
> From my understanding, a simple solution is to run Andersen’s or flow-sensitive analysis. For each objects you collected, get their reverse points-to (getRevPts), then you will get all pointers that access this object. You may only be interested in pointers are operants at stores, loads and getelementptrs.

Thank you for your answer,

I just tried it and it seems that if I have something like what I show, I end up with this llvm
```llvm
  %1 = alloca %struct.thing_s*, align 8 ;; allocation of p
  %4 = call noalias i8* @malloc(i64 noundef 16) #3
  %5 = bitcast i8* %4 to %struct.thing_s* ;; assignment of the result of malloc to p
```

But when I write to the field of the struct, llvm generates two instruction
```llvm
  %6 = load %struct.thing_s*, %struct.thing_s** %1, align 8 ;; get pointer to the struct pointer
  %7 = getelementptr inbounds %struct.thing_s, %struct.thing_s* %6, i32 0, i32 0
```

and getRevPts of the allocation, the malloc call instruction, only gives me the instruction that gets the pointer to the pointer not the GEP. Do you think exploring the VFG would help me with this?

**Comment by yuleisui** [comment]
For each getelementptr gep 
    Get the pointer p of gep
    Pts = getPts(p)
    If (pts.contains(o))
       this is the gep you want

**Comment by yuleisui** [comment]
> For each getelementptr gep
> Get the pointer p of gep
> Pts = getPts(p)
> If (pts.contains(o))
> this is the gep you want

Try this. I am typing through my phone, but the code should be self-explanatory

**Comment by enzo-pellegrini** [comment]
Ok thank you for your help, and have a nice day. I will consider your suggestion and maybe also try the VFG route as it sounds more efficient on a large program (the expectation is that I'm interested in a small minority of structs in the program, and doing a pointer analysis for all pointers could be worse than propagating through vfg, of course that is if I use a points-to analysis that doesn't do the analysis for all pointers in the program

**Comment by yuleisui** [comment]
Building SVFG needs a pointer analysis. Using pointer analysis is relatively cheaper. Feel free to explore other ways.
